packages/models-library/src/models_library/projects_nodes_io.py

```diff
@@ -6,22 +6,43 @@
         - Link to another port: PortLink
 """
 
+import re
+from enum import IntEnum
 from pathlib import Path
-from typing import Optional, Union
+from typing import Literal, Optional, Pattern, Union
 from uuid import UUID
 
-from pydantic import AnyUrl, BaseModel, Extra, Field, constr, validator
+from pydantic import AnyUrl, BaseModel, ConstrainedStr, Extra, Field, validator
 
+from .basic_regex import DATCORE_FILE_ID_RE, SIMCORE_S3_FILE_ID_RE, UUID_RE
 from .services import PROPERTY_KEY_RE
 
 NodeID = UUID
 
-# Pydantic does not support exporting a jsonschema with Dict keys being something else than a str
-# this is a regex for having uuids of type: 8-4-4-4-12 digits
-UUID_REGEX = (
-    r"^[0-9a-fA-F]{8}-?[0-9a-fA-F]{4}-?[0-9a-fA-F]{4}-?[0-9a-fA-F]{4}-?[0-9a-fA-F]{12}$"
+
+class NodeIDStr(ConstrainedStr):
+    regex: Optional[Pattern[str]] = re.compile(UUID_RE)
+
+
+# NOTE: this trick is used to keep backward compatility simcore.s3 is not a valid python variable name
+Location = IntEnum(
+    value="Location",
+    names=[("simcore.s3", 0), ("SIMCORE_S3", 0), ("datcore", 1), ("DATCORE", 1)],
 )
-NodeIDStr = constr(regex=UUID_REGEX)
+
+LocationID = Union[Literal[0], Literal[1]]
+LocationName = Union[Literal["simcore.s3"], Literal["datcore"]]
+
+
+class SimcoreS3FileID(ConstrainedStr):
```

Comment:
THOUGHT: All these are logic but I find them too convoluted for only two providers. 
I wonder how maintainable will be if we have more storage providers (e.g. dropbox, googledrive, ...)

Reply:
well I guess I would then separate in more files. cause each specific (S3, samba, etc..) have their own issues...